### ISSUE TITLE
Questions on the GAT conv layer

### ISSUE THREAD

**Issue by ldv1** [rusty1s/pytorch_geometric#1851]
## ❓ Questions & Help

I have a few questions from a newbie in PyTorch Geometric regarding the GAT model:

1/ In the `forward` method, we call `propagate` as follows:

` out = self.propagate(edge_index, x=(x_l, x_r), alpha=(alpha_l, alpha_r), size=size)`

Should I understand that `x_l` will map to the source features (`x_j` in `message`) whereas `x_r` will map to the target features (`x_i` in `message`) ? same thing for `alpha` ?

2/ In `message`, we treat the messages on all edges of the graph (well, the graphs in the mini-batch); I have hard times understanding how we can perform a softmax here; it is as if the `message` method was dealing only with the messages to a single target (can I stick to this view though it might not be right?). How is it done under the hood?

3/ What is the difference between a `Tensor` and a `OptTensor` ?

Thanks !

**Comment by rusty1s** [comment]
Hi,
1. That's correct. We refer to `x_l` as the source nodes, and `x_r` as the target nodes (in a bipartite graph). So for example, `GATConv((128, 256), 256)` will aggregate 128-dimensional feature vectors and combine them to 256-dimensional feature vectors.
2. Usually, `message` can be seen as a method that operates independently for each edge. However, in `GATConv` messages are inter-dependent (via the softmax). PyTorch Geometric provides a `softmax` function (`torch_geometric.utils.softmax`) that normalizes inputs across the same target nodes. This function is used here.
3. `OptTensor` describes an optional tensor, i.e. a `Tensor` or `None`. This might be useful in same cases, e.g., when you have a bipartite graph for which features only exist for source nodes (and you want to obtain target node features based on aggregating incoming edges). You can then do something like: `conv((x_l, None), edge_index)`.

**Comment by ldv1** [comment]
Thanks for your prompt reply.
